packages/state-transition/src/block/processWithdrawalRequest.ts

```diff
@@ -76,6 +76,7 @@ export function processWithdrawalRequest(
       withdrawableEpoch,
     });
     state.pendingPartialWithdrawals.push(pendingPartialWithdrawal);
+    state.pendingPartialWithdrawals.commit();
```

Comment:
can we commit once across multiple withdrawal requests?
eg:
```ts
for (...) {
  processWithdrawalRequest(...);
}
state.pendingPartialWithdrawals.commit();
```

Reply:
the problem is that we need to commit before we call `getPendingBalanceToWithdraw`

https://github.com/ChainSafe/lodestar/blob/06831cfc79ab667f4ce4ee897320ce3f4a495854/packages/state-transition/src/block/processWithdrawalRequest.ts#L46

since this calls `getAllReadonly`

https://github.com/ChainSafe/lodestar/blob/06831cfc79ab667f4ce4ee897320ce3f4a495854/packages/state-transition/src/util/validator.ts#L87

I am sure there is a better solution for this

Reply:
so if we don't commit after we push the first withdrawal request the issue happens

Reply:
@nflaig  calling `commit()` at the middle of state transition is not great but this is a limitation at ssz side, we should be able to get all items without calling `commit()`

call `state.pendingPartialWithdrawals.getAllReadonly()` at the start of `processWithdrawalRequest`

then modify `getPendingBalanceToWithdraw`
```typescript
let total = 0;
for (let i = 0; i < state.pendingPartialWithdrawals.length; i++) {
  const item = state.pendingPartialWithdrawals.get(i);
  if (item.validatorIndex === validatorIndex) {
    total += Number(item.amount);
  }
  
  return total;
}
```

performance should be the same, a unit test should help

Reply:
> call state.pendingPartialWithdrawals.getAllReadonly() at the start of processWithdrawalRequest
I think this would result in the same error, we need to call it outside of this loop
https://github.com/ChainSafe/lodestar/blob/06831cfc79ab667f4ce4ee897320ce3f4a495854/packages/state-transition/src/block/processOperations.ts#L74-L76

or we add something like `cachedPendingPartialWithdrawals` which is just a nomral array and we use that instead to check pending balance to withdraw

Reply:
`or we add something like cachedPendingPartialWithdrawals which is just a nomral array and we use that instead to check pending balance to withdraw`

Yea I was thinking about something like this. And whenever `state.pendingPartialWithdrawals` gets updated during state transition, the cache needs to be updated manually as well.

Reply:
something like https://github.com/ChainSafe/lodestar/pull/7375/commits/582ead563ff684e2812139bd8bfe04f40415e966 or you had somehting else in mind?